apps/aecore/src/aec_peer_connection.erl

```diff
@@ -0,0 +1,825 @@
+        {ok, Pid} ->
+            gen_tcp:controlling_process(TcpSock, Pid),
+            gen_server:cast(Pid, give_tcp_socket_ownership),
+            {ok, Pid};
+        Res ->
+            Res
+    end.
+
+retry(PeerCon) ->
+    gen_server:cast(PeerCon, retry).
+
+ping(PeerId) ->
+    call(PeerId, ping).
+
+get_header_by_hash(PeerId, Hash) ->
+    call(PeerId, {get_header_by_hash, Hash}).
+
+get_header_by_height(PeerId, Hash) ->
+    call(PeerId, {get_header_by_height, Hash}).
+
+get_n_successors(PeerId, Hash, N) ->
+    call(PeerId, {get_n_successors, Hash, N}).
+
+get_block(PeerId, Hash) ->
+    call(PeerId, {get_block, Hash}).
+
+get_mempool(PeerId) ->
+    call(PeerId, get_mempool).
+
+send_tx(PeerId, Tx) ->
+    call(PeerId, {send_tx, Tx}).
+
+send_block(PeerId, Tx) ->
+    call(PeerId, {send_block, Tx}).
+
+stop(PeerCon) ->
+    gen_server:cast(PeerCon, stop).
+
+
+call(PeerCon, Call) when is_pid(PeerCon) ->
+    gen_server:call(PeerCon, Call);
+call(PeerId, Call) when is_binary(PeerId) ->
+    case aec_peers:get_connection(PeerId) of
+        {ok, PeerCon}    ->
+            case PeerCon of
+                {connected, Pid} -> call(Pid, Call);
+                _ -> {error, no_connection}
+            end;
+        Err = {error, _} ->
+            Err
+    end.
+
+
+%% -- gen_server callbacks ---------------------------------------------------
+init([Opts]) ->
+    Version = <<?PROTOCOL_VERSION:16>>,
+    Genesis = aec_chain:genesis_hash(),
+    {ok, Opts#{ version => Version, genesis => Genesis }, 0}.
+
+handle_call(ping, From, State) ->
+    send_ping(State, From, local_ping_obj(State));
+handle_call({get_header_by_hash, Hash}, From, State) ->
+    send_get_header_by_hash(State, From, Hash);
+handle_call({get_header_by_height, Hash}, From, State) ->
+    send_get_header_by_height(State, From, Hash);
+handle_call({get_n_successors, Hash, N}, From, State) ->
+    send_get_n_successors(State, From, Hash, N);
+handle_call({get_block, Hash}, From, State) ->
+    send_get_block(State, From, Hash);
+handle_call(get_mempool, From, State) ->
+    send_get_mempool(State, From);
+handle_call({send_tx, Hash}, From, State) ->
+    send_send_tx(State, From, Hash);
+handle_call({send_block, Hash}, From, State) ->
+    send_send_block(State, From, Hash).
+
+handle_cast(retry, S = #{ host := Host, port := Port, status := error }) ->
+    Self = self(),
+    Pid = spawn(fun() -> try_connect(Self, Host, Port, 1000) end),
+    {noreply, S#{ status => {connecting, Pid} }};
+handle_cast(give_tcp_socket_ownership, S0 = #{ role := responder, tcp_sock := TcpSock }) ->
+    S = #{ version := Version, genesis := Genesis,
+           seckey := SecKey, pubkey := PubKey } = ensure_genesis(S0),
+    NoiseOpts = [ {noise, <<"Noise_XK_25519_ChaChaPoly_BLAKE2b">>}
+                , {s, enoise_keypair:new(dh25519, SecKey, PubKey)}
+                , {prologue, <<Version/binary, Genesis/binary>>}
```

Comment:
I understand that if different prologue than handshake fails. This is passing genesis hash - that makes sense (and BTW contains the initial consensus protocol version) - and *current* consensus/block protocol version - that I do not understand fully. What I would expect here is a peer-to-peer protocol specific version, that changes completely independently from the consensus protocol.

Reply:
Yes of course, throwing PROTOCOL_VERSION in there was just a hack that I forgot to fix this has little to do with consensus! Good catch!